inference/conflict.go

```diff
@@ -19,148 +19,177 @@ import (
+}
 
-// groupKey stores the type of the key representing a conflict group.
-// (Note that the type is currently set to any since the producer types of singleAssertionConflict and overconstrainedConflict
-// are not the same. Consolidating the two types of conflicts will help solve this problem (tracked in .)
-type groupKey any
+type nilFlow struct {
+	nilPath    []node // stores nil path of the flow from nilable source to conflict point
+	nonnilPath []node // stores non-nil path of the flow from conflict point to dereference point
+}
 
-// conflictGrouping stores groups of conflicts represented by their keys in a map. It also provides methods to add conflicts
-// to the groups and to generate diagnostics from the groups.
-type conflictGrouping struct {
-	groupMap map[groupKey][]conflict
+type node struct {
+	position token.Position
+	reason   string
 }
 
-// newConflictGrouping constructs and returns a pointer to conflictGrouping.
-func newConflictGrouping() *conflictGrouping {
-	return &conflictGrouping{
-		groupMap: make(map[groupKey][]conflict),
+func newNode(p annotation.Prestring) node {
+	nodeObj := node{}
+	if l, ok := p.(annotation.LocatedPrestring); ok {
+		nodeObj.position = l.Location
+		nodeObj.reason = l.Contained.String()
+	} else if p != nil {
+		nodeObj.reason = p.String()
 	}
+	return nodeObj
 }
 
-// addConflict adds a conflict to a conflict group identified by the conflict's key.
-func (t *conflictGrouping) addConflict(conflict conflict) {
-	key := conflict.key()
-	t.groupMap[key] = append(t.groupMap[key], conflict)
+func (n *node) String() string {
+	posStr := "<no pos info>"
+	reasonStr := "<no reason info>"
+	if n.position.IsValid() {
+		posStr = n.position.String()
+	}
+	if len(n.reason) > 0 {
+		reasonStr = n.reason
+	}
+	return fmt.Sprintf("\n\t-> %s: %s", posStr, reasonStr)
 }
 
-// diagnostics converts conflict groups to a list of analysis.Diagnostic for reporting.
-func (t *conflictGrouping) diagnostics() []analysis.Diagnostic {
-	var diagnostics []analysis.Diagnostic
-	for _, conflicts := range t.groupMap {
-		site := conflicts[0].getSiteMessage()
-		producer := conflicts[0].getProducerMessage()
-		consumer := ""
-		for _, c := range conflicts {
-			consumer += c.getConsumerMessage()
-		}
+// addNilPathNode adds a new node to the nil path.
+// LocatedPrestring contains accurate information about the position and the reason why NilAway deemed that position
+// to be nilable. We use it if available, else we use the raw string representation available from the Prestring `p`.
+func (n *nilFlow) addNilPathNode(p annotation.Prestring) {
+	nodeObj := newNode(p)
 
-		diagnostics = append(diagnostics, analysis.Diagnostic{
-			Pos:     conflicts[0].getPosition(),
-			Message: fmt.Sprintf("%s%s%s", site, producer, consumer),
-		})
-	}
-	return diagnostics
+	// Note that we prepend the new node to nilPath because we want to print the flow in the order in which it was discovered.
```

Comment:
This is missing some info, which I think is that these nodes get added starting from the conflict node, so they are added in reverse order from what the program flow would look like.

Reply:
Your are right. Modified the comment to explain this better.